### ISSUE TITLE
Investigate impact of long blocking abci handlers on p2p network

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#10075]
## What is the Problem Being Solved?

Under load, the cosmos and cosmic-swingset layers may block for a significant amount of time when processing blocks, in particular in the EndBlock (where swingset actually executes messages), and Commit (when the DB changes are large) phases.

It has been documented such loads negatively impact the performance of the chain. One of the earliest report is #5507, but more recently there has been numerous block production slowdowns where the chain appears to stall.

The surprising part is that on these slowdowns, our follower is able to execute and commit relatively quickly compared to the consensus block time: it might take 30s locally, yet the chain won't produce a new block for over 2 minutes. In theory once 67% of the voting power has processed the block, the chain should be able to make forward progress, however such a discrepancy in timing indicate something else may be at play here.

Furthermore, I have observed that in our instagoric networks, the non-primary nodes are not capable of making progress beyond the primary validator. This is possibly due to the these nodes not having direct p2p connectivity to the rest of the network, but instead being connected through the primary which is the one with public network connectivity.

There are a series of issues linked from https://github.com/cometbft/cometbft/issues/3245 that seem to imply that the p2p layer is dependent on the layers above not blocking.

## Description of the Design

Setup a 3 node chain as follow:
- 2 large / overprovisioned nodes A & C, which together have >= 67% of the voting power
- 1 resource constrained node B
- A & C do not have direct connectivity, and can only connect to B
- Place some load on the chain (real or synthetic)

We should verify that A & C clearly commit/vote on the block much faster than B. In that case we want to observe whether the chain makes forward progress as soon as A&C complete their block, or if B is somehow in the critical path.

## Security Considerations

None, investigation only

## Scaling Considerations

The chain should be able to make forward progress without slower nodes impeding as long as it has 67% of the voting power, regardless of the p2p topology of the nodes forming that voting power.

## Test Plan

See above

## Upgrade Considerations

None

**Comment by warner** [comment]
There is an environment variable `END_BLOCK_SPIN_MS` that will tell agoric-sdk to do a 100% busy-wait loop for that duration.

So the test should add that env var to the "B" (middle) node, and see if it causes p2p connectivity problems between A and C.

**Comment by toliaqat** [comment]
@Muneeb147 please help figure our p2p connectivity is causing issues or not, and then JC can help further investigate.

**Comment by Muneeb147** [comment]
So here's detailed analysis:

Used kubernetes for creating required setup where we had three nodes (validator-0, validator-1 and validator-2) which are directly connected (or peer dependent) on validator-primary node.

- The three nodes have in total of > 67% voting power.
- validator-primary has busy-wait enabled of 15mins.

The three validator nodes didn't create/commit blocks as soon as validator-primary went in busy-wait state. Checked logs of validators and they were throwing peer connectivity error and was retrying building connection.
The `agd status` was showing constant height which later incremented after 15mins and hence incremented on all validators too.

After removing the busy-wait, everything went smooth and all nodes were working fine (commiting/vote etc..)

So in a nutshell, the busy-wait on primary node affected the other nodes with peer connectivity error and the chain didn't proceed/progress

**Comment by JeancarloBarrios** [comment]
Ok this is a good confirmation, Now there is something we can legit tackle on the cometBFT side to see how we can improve our network connectivity

**Comment by mhofman** [comment]
I think the next steps are to
- see if the issue reproduces with `--with-tendermint=false` and running Tendermint Core in a separate process
- reproduce this issue using a plain cosmos app on unmodified sdk / cometbft, but otherwise using same version as we're currently on
  - this should actually include the ability to sleep in both EndBlock and in CommitBlock, and as such may require a small cosmos-sdk fork.
- upgrade to a bleeding edge version of the cosmos / cometbft and see if the problem still reproduces
